src/MerchantCenter/MerchantStatuses.php

```diff
@@ -652,11 +668,15 @@ public function process_product_statuses( array $statuses ): void {
 			}
 
 			$this->product_data_lookup[ $wc_product_id ] = [
-				'name'       => get_the_title( $wc_product ),
-				'visibility' => get_post_meta( $wc_product_id, $visibility_meta_key ),
-				'parent_id'  => $wc_product->post_parent,
+				'name'       => $wc_product->get_name(),
+				'visibility' => $wc_product->get_meta( $visibility_meta_key ),
+				'parent_id'  => $wc_product->get_parent_id(),
```

Comment:
I had the same feeling, the main reason I kept the` $product_data_lookup` as it was is because I initially intended to populate the product issues using the ProductView Report query. My plan was to reuse most of the code. However, after further investigation of the product issues, I realized that the ProductView doesn't return the details from the description or the documentation. So, I don't believe I'll be able to use the same query for that purpose.

https://github.com/woocommerce/google-listings-and-ads/blob/e9bc80d007c3574af4126949a67edc5519d4eb19/src/MerchantCenter/MerchantStatuses.php#L485-L487

One idea to retrieve the product issues could be to store the Google IDs when we retrieve the product status. Then, the product issues section could continue to function similarly to how it's currently implemented.

> Or if we don't want to save a copy in the class, can we have process_product_statuses call the other two functions internally so we don't need to call the three in a row? That way it can also pass the set of product objects by reference so we don't need to store it in the class instance.


Thanks for the suggestion! I've updated the code so that now those functions accept an array of products. While I was making these updates, I noticed that parent_id wasn't being fetched previously. So, I've adjusted the code to fetch all the parent IDs at once.

Reply:
> One idea to retrieve the product issues could be to store the Google IDs when we retrieve the product status. Then, the product issues section could continue to function similarly to how it's currently implemented.

Even though we can't use the ProductView Report query, we can still use the [productstatuses/list](https://developers.google.com/shopping-content/reference/rest/v2.1/productstatuses/list) endpoint, which supports pagination. That will allow us to pull the products in batches without needing to know a list of product/Google ID's up front. We can just act on whatever products the MC presents us with, and if we can map to a WC ID then we store the list of issues.

Since product issues are stored in the DB locally (until they expire), and we have the product statistics job run in batches of products. We can also consider doing something like this for each batch of UpdateMerchantProductStatuses:

- Query ProductView report to get a list of products
- Map products to WC ID's
- Add up intermediate statuses
- Fetch product statuses for the same batch of products using [productstatuses/custombatch](https://developers.google.com/shopping-content/reference/rest/v2.1/productstatuses/custombatch)
- Store issues in local DB

We might still need something to pull in account issues / add local issues to the table. And possibly decrease batch sizes if it takes more resources to pull in statuses and issues for that batch.

I think something like that will take the most advantage of running a background process, otherwise we'd have to create two to loop through the same batch of MC products. Let me know what you think about that approach.